qiskit_experiments/framework/experiment_data.py

```diff
@@ -767,6 +770,49 @@ def add_data(
     ) -> None:
         """Add experiment data.
 
+        Args:
+            data: Experiment data to add. Several types are accepted for convenience:
+
+                * Result: Add data from this ``Result`` object.
+                * List[Result]: Add data from the ``Result`` objects.
+                * Dict: Add this data.
+                * List[Dict]: Add this list of data.
+
+        Raises:
+            TypeError: If the input data type is invalid.
+        """
+
+        if any(not future.done() for future in self._analysis_futures.values()):
+            LOG.warning(
+                "Not all analysis has finished running. Adding new data may "
+                "create unexpected analysis results."
+            )
+        if not isinstance(data, list):
+            data = [data]
+
+        # Directly add non-job data
+        with self._result_data.lock:
+            for datum in data:
+                if isinstance(datum, dict):
+                    if datum["metadata"].get("composite_metadata"):
+                        tmp_exp_data = ExperimentData()
```

Comment:
You need another conditional branching here. Note that each element in the list contains the result of composite experiment. Let's assume a nested composite experiment consisting of three inner experiments. Here the `result_expX_circY` indicates a result of Y-th circuit of experiment X. The list of dictionary you may receive would look like:
```
[
    {result_expA_circ0, result_expB_circ0},  # loop 0
    {result_expC_circ0},  # loop 1
    {result_expA_circ1, result_expB_circ1},  # loop 2
    {result_expC_circ1},  # loop 3
]
```
Experiment A and B are batched by `ParallelExperiment` in this example. Now you need to create an `ExperimentData` instance with three child `ExperimentData` instance for A, B, C, and each child instance includes two `.data()` for circuit0 and circuit 1. What currently you are implementing is _making child instance for each loop_.

Instead, in each loop, you need to check whether corresponding child container is already initialized, make new one if not exist otherwise just call existing one, then add marginalize data to corresponding container. More specifically,
```
loop1: 
  - Create sub-container for A and B and set them to outer container
  - Marginalize count for A and B, and add marginalized count data of circuit0 to sub-container
loop2:
  - Create sub-container for C and set this to outer container
  - Skip marginalize because number of sub element is 1, add data to above
loop3:
  - Call child data for A and B
  - Marginalize and add data of circuit 1 to above
loop4:
  - Call child data for C
  - Add count data to above
```